src/classes/ContactsInLegacyHouseholds.cls

```diff
@@ -0,0 +1,594 @@
+        return [
+                SELECT Id, npo02__Household__c
+                FROM Contact
+                WHERE npo02__Household__c IN :householdCountMap.keyset()
+        ];
+    }
+
+    private void updateOldHouseholdNames(List<Contact> cons, Map<Id, Integer> householdCountMap) {
+        List<Id> nonEmptyHouseholdIds = getNonEmptyHouseholdIds(cons, householdCountMap);
+        if (nonEmptyHouseholdIds.size() > 0) {
+            updateHouseholdNameAndMemberCount(nonEmptyHouseholdIds);
+        }
+    }
+
+    private void deleteEmptyHouseholds(List<Contact> cons, Map<Id, Integer> householdCountMap) {
+        List<npo02__Household__c> emptyHouseholds = getEmptyHouseholds(cons, householdCountMap);
+        delete emptyHouseholds;
+    }
+
+    private List<Id> getNonEmptyHouseholdIds(List<Contact> cons,
+            Map<Id, Integer> householdCountMap) {
+        List<id> householdIdsNeedingNameUpdates = new List<id>();
+        for (Contact con : cons) {
+            if (!householdIsEmpty(householdCountMap, con)) {
+                householdIdsNeedingNameUpdates.add(householdIdFor(con));
+            }
+        }
+        return householdIdsNeedingNameUpdates;
+    }
+
+    private List<npo02__Household__c> getEmptyHouseholds(List<Contact> cons,
+            Map<Id, Integer> householdCountMap
+    ) {
+        List<npo02__Household__c> emptyHouseholds = new List<SObject>();
+        for (Contact con : cons) {
+            if (householdIsEmpty(householdCountMap, con)) {
+                emptyHouseholds.add(new npo02__Household__c(Id = householdIdFor(con)));
+            }
+        }
+        return emptyHouseholds;
+    }
+
+    private Boolean householdIsEmpty(Map<Id, Integer> householdCountMap, Contact con) {
+        return householdCountMap.get(householdIdFor(con)) < 1;
+    }
+
+    private Id householdIdFor(Contact con) {
+        return con.npo02__Household__c;
+    }
+
+    private void updateHouseholdNameAndMemberCount(List<Id> contactsNeedingHouseholdNamesUpdated) {
+        if (isFutureEligible()) {
+            HouseholdNamingService.updateHouseholdNameAndMemberCountAsynchronously(
+                    contactsNeedingHouseholdNamesUpdated);
+        } else {
+            HouseholdNamingService householdNamingService = new HouseholdNamingService();
+            householdNamingService.updateHouseholdNameAndMemberCount(contactsNeedingHouseholdNamesUpdated);
+        }
+    }
+
+    private Boolean isFutureEligible() {
+        return settings.isAsyncEnabled()
+                && !System.isFuture() && !System.isBatch();
+    }
+
+    public void onAfterUpdate() {
+        insertHouseholds();
+        fireRollups();
+        updateHouseholdNames();
+        deleteOldEmptyHouseholds();
+    }
+
+    private void deleteOldEmptyHouseholds() {
+        delete getEmptyHouseholdsToDelete(contactsThatChangedHouseholds());
+    }
+
+    private static Integer timesRun = 0;
```

Comment:
Yes, good eye - this is one thing I added during this refactor work.  This method runs recursively on merges and began hitting soql limits with the refactored work.  It is only used for this method and only to prevent rampant recursion in that merge case. I don't think its worth moving to another class, I think it fits best right here where it is the only place it is used.  I could rename it, but I think when you look at its usage in the method right below it is pretty clear.  But happy to rename if you feel strongly and have a name suggestion! :)

Reply:
Thanks for the explanation. For now, maybe rename to `updateHouseholdNamesCounter` and add a comment block above to explain why it's needed to prevent gov limit errors due to recursion during merges.